GH-91636: Clear weakrefs created by finalizers.

nascheme opened the pull request:
Weakrefs to unreachable garbage that are created during running of finalizers need to be cleared.  This avoids exposing objects that have `tp_clear` called on them to Python-level code.

Note that this is part of the GH-136189 change, split out to make it easier for back-porting.  This change is conceptually quite simple and is a candidate to backport to all older Python versions (this bug existed for a *long* time, perhaps in version 2.3).  There is some duplicated code between `handle_finalizers` and `clear_finalizers` but I think it's simpler just to leave that duplication.  In the "main" branch I'll clean that up.

<!-- gh-issue-number: gh-91636 -->
* Issue: gh-91636
<!-- /gh-issue-number -->

nascheme committed:
Clear weakrefs created by finalizers.

Weakrefs to unreachable garbage that are created during running of
finalizers need to be cleared.  This avoids exposing objects that
have `tp_clear` called on them to Python-level code.

colesbury commented on Python/gc_free_threading.c:
```diff
@@ -1494,7 +1494,7 @@ move_legacy_finalizer_reachable(struct collection_state *state)
 // enqueued in `wrcb_to_call`, but not invoked yet.
 static void
-clear_weakrefs(struct collection_state *state)
+handle_weakrefs(struct collection_state *state)
```
This is nearly identical to `clear_weakrefs`. I was a bit confused by the difference in naming. It would be clearer to me if there were just one function:

```c
clear_weakrefs(struct collection_state *state, bool enqueue_callbacks)
```

Where the first invocation we clear and enqueue and the second invocation we just clear the weakrefs.

nascheme replied:
Okay.

nascheme committed:
Reduce code duplication.

Add a boolean flag to indicate that callbacks should be enabled.  This
allows the same function to be used before and after finalizer
execution.

sergey-miryanov commented on Python/gc.c:
```diff
@@ -987,6 +994,10 @@ handle_weakrefs(PyGC_Head *unreachable, PyGC_Head *old)
+    if (!allow_callbacks) {
+        return 0;
+    }
+
```
Maybe it is worth to add an assert here for `old != NULL`?

neonene replied:
Similar question: The `old` and `allow_callbacks` arguments are orthogonal even in older versions?
